File: front/vellipsis/link.js
import { node } from "./node.js";


const link = (link_view, idx, link_size, node_view, node_size) => {
    const i = idx * link_size;
    return {
        idx: idx,
        l: link_view.getFloat64(i, true),
        s: link_view.getFloat64(i+8, true),
        damping: link_view.getFloat64(i+8*2, true),
        stress: link_view.getFloat64(i+8*3, true),
        a: node(node_view, link_view.getUint32(i+8*4, true), node_size),
        b: node(node_view, link_view.getUint32(i+8*4+4, true), node_size),
        uid: link_view.getUint32(i+8*5, true),
        active: link_view.getUint32(i+8*5+4, true),
    }
}

export {
    link,
}

File: engine/src/link.rs
use serde::{Deserialize, Serialize};

pub const LINK_SIZE: usize = 5 * 8 + 6 * 4;

#[derive(Deserialize, Serialize, Clone)]
pub struct Linking {
    pub stress_limit: f64,
    pub damping: f64,
    pub length: f64,
    pub strength: f64,
}

pub struct Link {
    pub length: f64,
    pub strength: f64,
    pub damping: f64,
    pub stress: f64,
    pub stress_limit: f64,
    pub a: usize,
    pub b: usize,
    pub uid: usize,
    pub idx: usize,
    pub active: u8,
}
